doc: Coding Style: Document cond. compilation in headers

carlescufi opened the pull request:
Document avoiding conditional compilation in headers.

Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>

galak reviewed (dismissed):
This should be in doc/guides/coding_guidelines/index.rst

carlescufi commented:
> This should be in doc/guides/coding_guidelines/index.rst

This is currently not included in the published documentation, and it has a table that has a MISRA column in it. I suggest we keep this here until the coding guidelines are actually published.

carlescufi commented:
> You may conditionally exclude fields within structure definitions to avoid wasting memory when the feature they support is not enabled.

I like the proposal, adopted it thanks.

pabigot commented on doc/contribute/index.rst:
```diff
@@ -356,7 +356,12 @@ following exceptions:
+  * Do not conditionally compile function declarations in header files.
+  * Do not conditionally compile structure declarations in header files.
+  * You may conditionally exclude fields within structure definitions to avoid
+    wasting memory when the feature they support is not enabled.
```
This will need clarification if/when we get `DOXYGEN` added to `doc/zephyr.doxyfile.in` replacing `DT_DOXYGEN`, because the guideline should then explain *how* to conditionally exclude those fields.

nashif replied:
the DOXYGEN macro might work for some cases, but I really do not want to see code where Kconfig variables are set to values in headers files to satisfy doxygen, this is going to be a lot of noise and confusing above all.

galak commented:
> This is currently not included in the published documentation, and it has a table that has a MISRA column in it. I suggest we keep this here until the coding guidelines are actually published.

Is there a reason its not published?

nashif commented:
> Is there a reason its not published?

it is already in the tree, job that publish that will run at some point or you can poke it...

nashif commented:
> This should be in doc/guides/coding_guidelines/index.rst

I would not add this in the page, but under the same index and TOC maybe, this will become a really lengthy table..

galak commented:
> it is already in the tree, job that publish that will run at some point or you can poke it...

Confused, this looks like it was merged some time ago, so why hasn't shown up in the daily docs build?

galak commented:
> Confused, this looks like it was merged some time ago, so why hasn't shown up in the daily docs build?

Never mind, @pabigot pointed it out to me -- github shows the author date.

carlescufi commented:
@galak and @nashif we clearly have a disagreement here: Either we have a single list of Coding Guidelines (which is what @galak proposed during the meeting) or we have multiple lists. The current table seems ill suited for non-MISRA rules:
https://docs.zephyrproject.org/latest/guides/coding_guidelines/index.html

So my proposal would be to have a separate table on that same page.

galak commented:
> @galak and @nashif we clearly have a disagreement here: Either we have a single list of Coding Guidelines (which is what @galak proposed during the meeting) or we have multiple lists. The current table seems ill suited for non-MISRA rules:
> https://docs.zephyrproject.org/latest/guides/coding_guidelines/index.html
> 
> So my proposal would be to have a separate table on that same page.

I'm good with that.  Mine main point is this should be part of the coding guidelines (and not Style).  How we document it I'm not as concerned about.

nashif commented:
> The current table seems ill suited for non-MISRA rules:
> [docs.zephyrproject.org/latest/guides/coding_guidelines/index.html](https://docs.zephyrproject.org/latest/guides/coding_guidelines/index.html)

agree with that, all guidelines should be in the same section of the documentation, same page or a subpage that is something that will be judged by usability and how large those tables will grow...

Note that the MISRA table is autogenerated, so I do not want to add anything manually to it..

MaureenHelm commented:
> agree with that, all guidelines should be in the same section of the documentation, same page or a subpage that is something that will be judged by usability and how large those tables will grow...

Agree

carlescufi committed:
doc: guidelines: Cleanup formatting

Cleanup formatting and a couple of instances of missing "rules".

Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>

carlescufi committed:
doc: guidelines: Add additional guidelines

Aside from the MISRA-C derived Coding Guidelines, add a list for
additional ones that are not a subset of a standard.

Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>

carlescufi committed:
doc: Link to guidelines from the contribution page

Link to the coding guidelines from the main contribution guidelines
page, so that users are aware of the guidelines.

Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>

carlescufi committed:
doc: contribution: Add link to the coding guidelines

Add a small section that describes the existance of the coding
guidelines page and links to it, in order to raise awareness of the
effort towards complying with those.

Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>

carlescufi commented:
@nashif @galak @MaureenHelm @pabigot please take a look

pabigot reviewed (approved):
Seems fine; looking at #25851 I was already almost compliant with this.

MaureenHelm commented on doc/contribute/index.rst:
```diff
@@ -249,6 +249,13 @@ The `builds@lists.zephyrproject.org mailing list
+
+Beyond the :ref:`coding_style` that Zephyr enforces for all code that is
+submitted for inclusion, the project targets compliance with a series of
+coding_guidelines. Refer to the :ref:`coding_guidelines` section of the
```
```suggestion
coding guidelines. Refer to the :ref:`coding_guidelines` section of the
```

carlescufi replied:
Done, thanks!

pabigot reviewed (changes requested):
Thinking this through a little more: I would really like to see these extra rules accompanied by a short rationale: why are we asking people to do this?

In this case it'd be something like: Excluding declarations from the header based on compile-time options may prevent their documentation from being generated.  Their absence also prevents use of `if (IS_ENABLED(CONFIG_FOO)) {}` as an alternative to preprocessor conditionals when the code path should change based on the selected options.

Can we work that in?

carlescufi commented:
> Thinking this through a little more: I would really like to see these extra rules accompanied by a short rationale: why are we asking people to do this?
> 
> In this case it'd be something like: Excluding declarations from the header based on compile-time options may prevent their documentation from being generated. Their absence also prevents use of `if (IS_ENABLED(CONFIG_FOO)) {}` as an alternative to preprocessor conditionals when the code path should change based on the selected options.
> 
> Can we work that in?

Since you've been so kind as to write the rationale for me, of course! :)

nashif commented on doc/guides/coding_guidelines/index.rst:
```diff
@@ -784,3 +787,26 @@ and the equivlent rules from other standards for reference.
+Additional rules
+================
+
+.. list-table:: Additional rules
```
does it really need to be a table? this is very limiting and would make it difficult to add examples and additional context etc, maybe a definition list or some other type of list? the MISRA-C rules are in the table because they are references, this does not need to be the same for additional rules that we full maintain...

carlescufi replied:
I thought about this, yes. It doesn't really need to be a table, it's just that I thought consistency with the other guidelines looked nice. But I agree, and while I was making the table I was asking myself the same question. I will rewrite as a section-separated list.

pabigot commented on doc/guides/coding_guidelines/index.rst:
```diff
@@ -784,3 +787,26 @@ and the equivlent rules from other standards for reference.
+         avoid wasting memory when the feature they support is not enabled. |br|
+      -  Excluding declarations from the header based on compile-time options
+         may prevent their documentation from being generated. Their absence
+         also prevents use of if ``(IS_ENABLED(CONFIG_FOO)) {}`` as an alternative
```
`if` needs to be inside the literal region.

carlescufi commented:
@pabigot @nashif @galak @MaureenHelm please take another look

nashif commented on doc/guides/coding_guidelines/index.rst:
```diff
@@ -784,3 +787,30 @@ and the equivlent rules from other standards for reference.
+Excluding declarations from the header based on compile-time options may prevent
+their documentation from being generated. Their absence also prevents use of
+``if (IS_ENABLED(CONFIG_FOO)) {}`` as an alternative to preprocessor
+conditionals when the code path should change based on the selected options.
```
Shouldn't there be other exceptions listed? Look at many of the header files now, for example CONFIG_USERSPACE conditionals are needed in many of the places.

ceolin replied:
Isn't this rule about conditionally compile functions and types in headers ? How this is related with `IS_ENABLED` that is used with generated symbols ?

pabigot replied:
@ceolin If the material enabled by `CONFIG_DEVICE_POWER_MANAGEMENT` were unconditionally declared in `<device.h>` rather than wrapped in `#ifdef CONFIG_DEVICE_POWER_MANAGEMENT` you could do:

    int rc = -ENOTSUP;
    struct device *cons = device_get_binding(DT_LABEL(DT_CHOSEN(zephyr_console)));

    if (IS_ENABLED(CONFIG_DEVICE_POWER_MANAGEMENT)) {
        rc = device_set_power_state(cons, DEVICE_PM_ACTIVE_STATE, NULL, NULL);
    }

rather than having to do:

    int rc = -ENOTSUP;
    struct device *cons = device_get_binding(DT_LABEL(DT_CHOSEN(zephyr_console)));

    #ifdef CONFIG_DEVICE_POWER_MANAGEMENT
    rc = device_set_power_state(cons, DEVICE_PM_ACTIVE_STATE, NULL, NULL);
    #endif

because the latter generates build errors when `device_set_power_state` and `DEVICE_PM_ACTIVE_STATE` are undefined as they are when `CONFIG_DEVICE_POWER_MANAGEMENT=n` currently.

Some developers find the former to be easier to understand.

ceolin replied:
if `CONFIG_DEVICE_POWER_MANAGEMENT` is undefined `device_set_power_state` won't be compile so no error should happen. Am I missing something ?

Regarding the macro `IS_ENABLED` it should no be encouraged, actually should be removed. It violates rule 14.3 `Controlling expressions shall not be invariant`.

pabigot replied:
> if `CONFIG_DEVICE_POWER_MANAGEMENT` is undefined `device_set_power_state` won't be compile so no error should happen. Am I missing something ?

I think so, but I can't really tell what.  Somebody else might be able to explain it better.

> Regarding the macro `IS_ENABLED` it should no be encouraged, actually should be removed. It violates rule 14.3 `Controlling expressions shall not be invariant`.

Heh.  OK.  Should be interesting to see how people respond to that one.

jhedberg replied:
> Should be interesting to see how people respond to that one.

Here it comes :)

> Some developers find the former to be easier to understand.

`IS_ENABLED()` isn't just about readability. Using #ifdefs means you increase the number of build-time configurations both the developer and CI needs to test in order to guarantee that everything builds fine. With `IS_ENABLED()` the compiler will always "see" the code and let you know of syntax errors, etc, even if the Kconfig option in question is not enabled.

With #ifdefs and CI, in the best case (when the developer remembered to add build tests for all configuration combinations) you've increased the load and and the time CI takes to complete (which is already a problem for us). In the worst case you've introduced code to the tree that's never build tested and possibly broken.

> Regarding the macro IS_ENABLED it should no be encouraged, actually should be removed. It violates rule 14.3 Controlling expressions shall not be invariant.

I'd argue that both the improved readability aspect and the reduced number of builds & configurations to verify syntactical correctness of the code contribute to improved code quality, and therefore any coding guideline requiring the opposite would be self defeating.

jhedberg replied:
> It violates rule 14.3 Controlling expressions shall not be invariant.

Just wanted to add to this that I don't think using IS_ENABLED() with Kconfig symbols is necessarily covered by that rule (or at least its spirit), since we're not talking about something that's completely invariant here. We might want to still have this rule but make it clear that it doesn't apply to values which may change based on build-time configuration.

ceolin replied:
> > Should be interesting to see how people respond to that one.
> 
> Here it comes :)
> 
> > Some developers find the former to be easier to understand.
> 
> `IS_ENABLED()` isn't just about readability. Using #ifdefs means you increase the number of build-time configurations both the developer and CI needs to test in order to guarantee that everything builds fine. With `IS_ENABLED()` the compiler will always "see" the code and let you know of syntax errors, etc, even if the Kconfig option in question is not enabled.

Ok, that I understand but that does not mean build errors when `CONFIG_DEVICE...` is undefined, that was the reason of my question :)

The unconditional syntax check is something that I can't argue against, it is a good thing.

> 
> With #ifdefs and CI, in the best case (when the developer remembered to add build tests for all configuration combinations) you've increased the load and and the time CI takes to complete (which is already a problem for us). In the worst case you've introduced code to the tree that's never build tested and possibly broken.
> 
> > Regarding the macro IS_ENABLED it should no be encouraged, actually should be removed. It violates rule 14.3 Controlling expressions shall not be invariant.
> 
> I'd argue that both the improved readability aspect and the reduced number of builds & configurations to verify syntactical correctness of the code contribute to improved code quality, and therefore any coding guideline requiring the opposite would be self defeating.

I don't completely agree with the readability argument, the unconditional syntax check is really something to consider. Regarding the rule itself we definitely can document this deviation or "reinterpretation". But keep in mind that this rule won't be the only one, this will cause other violations due unreachable code.
